### DISCUSSION TITLE
Support multi-termination cabling through fiber modules

### DISCUSSION THREAD

**Discussion by shepherdjay** [netbox-community/netbox#13955] [Ideas]
Per comments in #13930 this is what I want to try to model where there is a modular fiber patch panel:

[![](https://mermaid.ink/img/pako:eNqVU8tugzAQ_BVrzyESqOqBW6Vcc2pvOIf1A7BkbGSM0ijKv9fgQgOISuE03pnZnbXwHbgVEnIotb3yGp0nXydqSPi6q_K8Tp8P2QAXLEmOhIK2leKoKRyTSblSJRTOVvRakg_S2jAlpRCqCSkVky5tMOrHU9ZgoMieYzmgZ5XDtiZtfeuGDIQj08pUkZ57fnNrjOS-yN5I5x0aEeuXlS6dZqfZr657Djq12bjYa645drRvQxQR5Jgm79uMbKLZkpZG7Oz9X7xsuv0XXWxvn2zVLgwoKEQ4bkThspGwPwlbS-a9AoADNNI1qET4a-9DmYKvZSMp5AEKWWKvPQVqHkGKvbefN8Mh966XB-hbgV6eFIa4DeQl6i5UpVDeunN8CeODePwACx3w_w?type=png)](https://mermaid.live/edit#pako:eNqVU8tugzAQ_BVrzyESqOqBW6Vcc2pvOIf1A7BkbGSM0ijKv9fgQgOISuE03pnZnbXwHbgVEnIotb3yGp0nXydqSPi6q_K8Tp8P2QAXLEmOhIK2leKoKRyTSblSJRTOVvRakg_S2jAlpRCqCSkVky5tMOrHU9ZgoMieYzmgZ5XDtiZtfeuGDIQj08pUkZ57fnNrjOS-yN5I5x0aEeuXlS6dZqfZr657Djq12bjYa645drRvQxQR5Jgm79uMbKLZkpZG7Oz9X7xsuv0XXWxvn2zVLgwoKEQ4bkThspGwPwlbS-a9AoADNNI1qET4a-9DmYKvZSMp5AEKWWKvPQVqHkGKvbefN8Mh966XB-hbgV6eFIa4DeQl6i5UpVDeunN8CeODePwACx3w_w) 

I actually do think this might be achievable without going too much into strands of the cable. I'm having trouble fully understanding the code but it appears terminations are submitted and stored as lists which are ordered. 

So if you wanted to represent the above model you would create a cable like this:

```
A Side
Device: fiber1
terminations
a1-6 (positions 6), b1-6 (positions6)

B Side
Device: fiber2
terminations
a1-6 (positions 6), b1-6 (positions6)
```

And what if you needed to flip so that strands 13-24 connected to module A on fiber2 instead (basically having rolled the cable) you could reverse the order of the terminations 

```
A Side
Device: fiber1
terminations
a1-6 (positions 6), b1-6 (positions6)

B Side
Device: fiber2
terminations
 b1-6 (positions6), a1-6 (positions 6)
```

In the event that there are mismatched number of positions that is where a path split condition could occur. This type of logic could only apply to cables whose terminations have positions in the first place such as rear ports.

**Comment by jeremystretch** [comment]
This is an unrelated tangent, but could you share how you created the topology diagram above? It looks like Mermaid, maybe?

**Reply by shepherdjay** [reply]
Yes this was using mermaid -- specifically https://mermaid.live/ -- which you can export to markdown and you can create the diagram as a link. If you click on it it will take you straight to the mermaid code.

**Reply by DanSheps** [reply]
This is cool, bookmarked, thanks!

**Reply by DanSheps** [reply]
@shepherdjay could you give the code you used to generate the above?

**Reply by shepherdjay** [reply]
@DanSheps -- Mobile at the moment but the image itself should be a hyperlink that will also take you right to the code used to generate it.

**Reply by DanSheps** [reply]
You are correct, thanks!

**Comment by DanSheps** [comment]
Unfortunately, what you are requesting is going to logistically be a nightmare to model.

In current model, you have Front Ports, Rear Ports and cables (going to ignore interfaces because they are not relevant here).

Rear Ports have positions (1, 6, 12, whatever) but is still a single object, the positions are a virtual concept.

Front Ports are always linked to a rear port (and virtually to a specific position on the rear port).

When you cable from a rear-port to a rear port,  the positions are 1:1 mapped (position 1 on one side is position 1 on the other, 6 on one to 6 on the other).

Connecting ports, there is no concept of order of the ports (however typically it is ordered by PK), so if you have port a1 and b1 on one side a1 and b1 on another side of a cable, that doesn't mean a1 matches up with b1.  You are better to model the "buffer tube" (the 12 strands) in this instance instead of the 24 cables.

I do want to point out something, very important.  The "connecting multiple ports with 1 cable" was to solve the "duplex cable to interface" problem and really wasn't meant to connect two rear port sets together in this manner.  This isn't really someothing you can do in real life either, if you view a cable in netbox "properly".  A cable in netbox should be viewed as a single connenctor (simplex LC, duplex LC, MPO) and not multiple connectors bundled together (2 MPO-12 on a 24 strand fiber)

**Reply by DanSheps** [reply]
I put this in another issue:

> I think the big thing is going to be you shouldn't be using multi-position rear ports with multi-plex cables as we have no easy/clean way of telling which cable which rear port should connect to. You should instead use a single cable for each rear port (typically rear ports will be your plant fiber, if you are using multi-position ports, that is your "buffer" (12 fibers strands bundled). If you think of a 244, you would have:
>
> * 288 strands total
> * 24 buffers (Buffer tube really, Blue, Green, ... Blue-1, Green-1, etc)
> * 12 strands per buffer
>
> Rear ports, with 12 positions, should be a single cable that is your "buffer". You shouldn't try and connect 2 "buffers" to 2 rear-ports.

**Reply by shepherdjay** [reply]
> there is no concept of order of the ports (however typically it is ordered by PK), so if you have port a1 and b1 on one side a1 and b1 on another side of a cable, that doesn't mean a1 matches up with b1.

Today anyway is what you mean right? 

I'm not familiar with the code enough itself but an ordered list is certainly a doable concept in Python, Django, and SQL. The cable may not store terminations today in an ordered format but that is why this is being raised as an idea. My naive reading of the code implies it [does store terminations as an ordered list](https://github.com/netbox-community/netbox/blob/4286c1cde255a9bf146b3b192e1ac17566af0094/netbox/dcim/models/cables.py#L135) but could definitely be wrong here.

> ... This isn't really someothing you can do in real life either, if you view a cable in netbox "properly". 

I'm not 100% clear what you mean. I think, feel free to correct if wrong, what you are trying to say is that a cable in netbox should be modeled exclusively as a one-to-one connection. And while that may be how netbox started out as its clear the desire is there to model cables with multi-terminations. IE.. the desire is to model a cable in the real world... as a cable. We have one 24 strand cable of fiber... it is spliced into 2x modules that consume 12 strands each. So its not like I'm inventing use cases here, I'm simply trying to determine the best way to be able to accurately model that in some fiber panels we have 1x 24 strand cable running to the panel... and in others its 2x 12 strand cables.

Previously netbox allowed this connection type -- but would show path split error. It's only recently this broke completely (or fixed depending on pov) for this type of cable connection. I don't know if the logic is insurmountable in the cable trace logic.

But to be fair the cable path logic seems very complex approaching 200 loc for a single function so no doubt I am understating the difficulty of the problem.

**Reply by DanSheps** [reply]
> Today anyway is what you mean right?

I am talking about in the model now, correct.

> I'm not familiar with the code enough itself but an ordered list is certainly a doable concept in Python, Django, and SQL. The cable may not store terminations today in an ordered format but that is why this is being raised as an idea. My naive reading of the code implies it [does store terminations as an ordered list](https://github.com/netbox-community/netbox/blob/4286c1cde255a9bf146b3b192e1ac17566af0094/netbox/dcim/models/cables.py#L135) but could definitely be wrong here.

Certainly it is do-able, however your understanding of the code is not 100% correct.  The list is ordered, at that time, but that order is non-deterministic (that order can be influenced by outside variables, for example of the port names change, the order may change).  For multi-termination on rear ports to work, it needs to be a determinstic order, however you also run into problems here too with simply relying on something deterministic, for example where 1 port is deleted you have 24 on one side and 12 on another.

Another option would be to introduce a port->cable position mapping, however that would require an increase in the number off form fields/data storage on the cable model (or we might evevn need to go to another intermediate model to handle it), which just isn't feasible.

> I'm not 100% clear what you mean. I think, feel free to correct if wrong, what you are trying to say is that a cable in netbox should be modeled exclusively as a one-to-one connection. And while that may be how netbox started out as its clear the desire is there to model cables with multi-terminations. IE.. the desire is to model a cable in the real world... as a cable. We have one 24 strand cable of fiber... it is spliced into 2x modules that consume 12 strands each. So its not like I'm inventing use cases here, I'm simply trying to determine the best way to be able to accurately model that in some fiber panels we have 1x 24 strand cable running to the panel... and in others its 2x 12 strand cables.

What I am saying is NetBox should, as close as practicable, replicate real life.  With that goal stated, lets look at a couple of options for fiber plant:

1. 12 port LC to a 12 position MPO (6 duplex, example product: FC29N-12-10U https://www.panduit.com/en/products/fiber-optic-systems/fiber-optic-panels-cassettes-enclosures/fiber-optic-cassettes/fc29n-12-10u.html)
2. 24 port strand to 2x12 position MPO (12 duplex, example product: FC29N-24-10U https://www.panduit.com/en/products/fiber-optic-systems/fiber-optic-panels-cassettes-enclosures/fiber-optic-cassettes/fc29n-24-10u.html)
3. 24 port LC to 24 port LC Bulkhead 12 duplex, https://www.panduit.com/en/products/fiber-optic-systems/fiber-optic-panels-cassettes-enclosures/fiber-optic-panels/fap12wbudlcz.html)

There are likely multiple ways to model a back-to-back connection for this, but lets take a look at what is the closest real-world example(s) for each:

1. [FC29N-12-10U [MPO-1]] --Cable 1-- [[MPO-1]FC29N-12-10U]
2. [FC29N-24-10U [MPO-1]] --Cable 1-- [[MPO-1]FC29N-24-10U]
    [FC29N-24-10U [MPO-2]] --Cable 2-- [[MPO-2]FC29N-24-10U]
3. [FAP12WBUDLCZ [LC-1]] --Cable 1-- [[LC-1]FAP12WBUDLCZ] 
    [FAP12WBUDLCZ [LC-2]] --Cable 2-- [[LC-2]FAP12WBUDLCZ] 
    [FAP12WBUDLCZ [LC-3]] --Cable 3-- [[LC-3]FAP12WBUDLCZ] 
    [FAP12WBUDLCZ [LC-4]] --Cable 4-- [[LC-4]FAP12WBUDLCZ] 
    [FAP12WBUDLCZ [LC-5]] --Cable 5-- [[LC-5]FAP12WBUDLCZ] 
    [FAP12WBUDLCZ [LC-6]] --Cable 6-- [[LC-6]FAP12WBUDLCZ] 
    [FAP12WBUDLCZ [LC-7]] --Cable 7-- [[LC-7]FAP12WBUDLCZ] 
    [FAP12WBUDLCZ [LC-8]] --Cable 8-- [[LC-8]FAP12WBUDLCZ] 
    [FAP12WBUDLCZ [LC-9]] --Cable 9-- [[LC-9]FAP12WBUDLCZ] 
    [FAP12WBUDLCZ [LC-10]] --Cable 10-- [[LC-10]FAP12WBUDLCZ] 
    [FAP12WBUDLCZ [LC-11]] --Cable 11-- [[LC-11]FAP12WBUDLCZ] 
    [FAP12WBUDLCZ [LC-12]] --Cable 12-- [[LC-12]FAP12WBUDLCZ]

These are the "proper" ways to model them

You shouldn't use a multi-termination cable represent a distinct cable path, which is what the buffer tube on a 24 fiber interconnect with 2x12 MPO connectors on each end is.  The reason being is, in theory, those two MPO's on one end **can** go to a different cassette (I do this frequently).

You need to look at cables in NetBox in a similar manner in whichh you do in real life:

* You pick up a duplex cable, it has 2 positions, it can go into 2 LC ports (typically beside each other) so it should be modelled as 1 cable with 1 positions.
* You pick up a MPO cable, it has 1 MPO connector with 12 positions.  This should be 1 cable (it could go into 12 "MPO" front ports, or a single "MPO" rear port with 12 positions).
* You pick up a MPO cable, it has 2 MPO connectors with 12 positions each.  This should be modelled as 2 cables, 1 cable per MPO.

> We have one 24 strand cable of fiber... it is spliced into 2x modules that consume 12 strands each. 

Next time you get a chance, watch them splice it.  First off, strands will typically be spliced into a cassette or into an individual LC pigtail.  You will see there are typicially 2 buffer tubes in a single 24 strand jacket.  Each buffer tube will be color coded (generally) as well as each fiber strand will have a color code.  For reference:

Fiber Number | Color
-- | --
1 | Blue
2 | Orange
3 | Green
4 | Brown
5 | Slate
6 | White
7 | Red
8 | Black
9 | Yellow
10 | Violet
11 | Rose
12 | Aqua


When dealing with MPO-16 there are a few more colors but these 12 are typical for anything not "odd numbered".

> Previously netbox allowed this connection type -- but would show path split error. It's only recently this broke completely (or fixed depending on pov) for this type of cable connection. I don't know if the logic is insurmountable in the cable trace logic.

We probably shouldn't have

> But to be fair the cable path logic seems very complex approaching 200 loc for a single function so no doubt I am understating the difficulty of the problem.

It is very complex, believe me.  There are likely other methods we could use, but they would result in an exponential increase in processing time/power required.

There might be some options if we ever get to the point where we track all of the strands, buffer tubes, and jacket for eachh cable.  We have more important features we need to push out at this point in time though.

**Reply by shepherdjay** [reply]
I appreciate the very in-depth reply so I apologize in advance for my reply being short in comparison. My intent is not to dismiss your reply but I think on the whole it can be summarized very nicely with this statement.

> You need to look at cables in NetBox in a similar manner in whichh you do in real life:

That's exactly what I'm doing. I think we just don't look at cables the same way 😃. Just as I don't view the 8 separate copper connectors in a cat 6 cable as 8 distinct cable paths, I don't really view the 24 strands as 24 distinct cable paths (or 2)... I'm certainly not labeling the individual strands or buffer tubes. At the end of the day for us it's a single cable coming out of the back of the tray. It happens that the tray is modular so I was trying to represent that accurately.

Ultimately it sounds like this is just a trade-off. I can either represent the cable how I view it.. or the modular fiber tray how I view it... but not both. Likely the simplest tradeoff is to just drop the modular fiber tray and have a 6, 12, 18, 24 position rear versions of the same fiber tray depending on how they have been populated.

**Reply by DanSheps** [reply]
> That's exactly what I'm doing. I think we just don't look at cables the same way 😃. Just as I don't view the 8 separate copper connectors in a cat 6 cable as 8 distinct cable paths, I don't really view the 24 strands as 24 distinct cable paths (or 2)... I'm certainly not labeling the individual strands or buffer tubes. At the end of the day for us it's a single cable coming out of the back of the tray. It happens that the tray is modular so I was trying to represent that accurately.

The difference is the 8 strands are alll part of the same connector; with fiber you either have a single connector (LC-LC) for each strand or something like a MPO-4, MPO-8, MPO-12, MPO-16 going into a module.

> > I can either represent the cable how I view it.. or the modular fiber tray how I view it...

Why not both?  You don't mention the tray, but I am just going to assume something like a FAP-12 in 2 bays.  If it is a FAP-12 (or similar), you should ideally be modelling it as separate strands as they are separate connectors.  However if you want to go simpler, you can model the "buffer tubes", so instead of 24 cables you are down to 2.  You could even have a script to automate this for you.

> Ultimately it sounds like this is just a trade-off

There is definitely room for improvement, however the cabling model is unlikely to change in the near future.  But all the things I mentioned in my previous post could be better represented if the Cabling model was extended, however there is a trade-off of added complexity, both in code and in the manner in which you would connect things.

Take for example if we had a "cable", and that cable included the buffer tubes and strands.  You could easily have the buffer tubes connect  to a rear-port and then imply the position of the strand is the position of the cable, however what if at the other end you don't have a rear-port with multiple positions, you would then need to cable every single strand to a rear port.

While the cabling model right now is rather simplistic, it can somewhat accurately describe your cabling

**Reply by shepherdjay** [reply]
> Why not both?

Because its not possible to model it in the way I see it.. your next statement states why. That it is either 24 separate cables, 12 separate duplex pairs, or 2 separate buffer tubes (one for each module) -- None of those three items is how I see the cable. None of those is how I would label the cable. I understand this is different from the way you or even possibly most people see this. But when I look at something like this https://www.lanshack.com/24-Strand-Indoor-Interlocking-Armor-Plenum-1040100-GIG-50125-OM4-Multimode-Fiber-Optic-Cable-by-the-Foot-P8740 coming out of the back of a fiber tray... this is one cable to me.

So if I want to display the cable as one cable in the trace. I have a couple of choices.
- Drop the modularity of the fiber tray in favor of specific fiber tray fill outs with either 6, 12, 18, or 24 rear positions along with every other possible combination as it ends up existing in our environment.
- Create a standin adapter that takes 2 front ports to 1 rear port with 2 positions. Possibly this can be created as an internal reference utilizing front/rear ports internally connected instead of separate device.

**Reply by DanSheps** [reply]
If you want to give me specific part numbers for the tray as well as the cassettes/bulkheads used I can probably come up with a way to properly model it, however whatever method I propose, you aren't going to view the cable as a "single cable" with "24 strands".

> Drop the modularity of the fiber tray in favor of specific fiber tray fill outs with either 6, 12, 18, or 24 rear positions along with every other possible combination as it ends up existing in our environment.

I would argue, that mapping your buffer tubes as a cable to the 12 port rear position is *more* accurate then what you are doing now.  You state you have 2 modules within your Fiber tray, with 24 strands.  Guaranteed that there are 2 buffer tubes in the cable you specify.  Now, lets say that somewhere down the line you want to drop a closet in the middle of that cable.  You can mid-span that cable and pull out one of those buffer tubes and splice it into  the new tray, you now don't have 1 cable you have 2.  This is why you should at the bare minimum be using the buffer tubes as your "cables" instead of the whole 24/48/96/144/288 as a cable.

**Reply by shepherdjay** [reply]
> If you want to give me specific part numbers for the tray as well as the cassettes/bulkheads used I can probably come up with a way to properly model it, however whatever method I propose, you aren't going to view the cable as a "single cable" with "24 strands".

Ha - I've been having a hard time with that as well. Its a Corning Lanscape type system.

> I would argue, that mapping your buffer tubes as a cable to the 12 port rear position is more accurate then what you are doing now.

Almost certainly it is more accurate, just as modeling the 24 strands would be even more accurate. As would modeling the 8 copper strands on a cat6 would be more accurate in cases where those get converted to pots lines.

> Now, lets say that somewhere down the line you want to drop a closet in the middle of that cable. You can mid-span that cable and pull out one of those buffer tubes and splice it into the new tray, you now don't have 1 cable you have 2. This is why you should at the bare minimum be using the buffer tubes as your "cables" instead of the whole 24/48/96/144/288 as a cable.

Meh... I try not to add a bunch of work or complexity for things that might happen in the future. When we run the 24 strand its because there are already 2 modules we are populating. We don't run those 24 strand ahead of time so its highly unlikely we would ever disconnect one of those modules and midspan it vs just running another cable.

Either way I think this idea still has merit though perhaps not enough use to justify any work at this point in time.
